### ISSUE TITLE
[Feature Request] Add a cluster setting for memory threshold to pick OrdinalsCollector in Cardinality aggregation

### ISSUE THREAD

**Issue by rishabhmaurya** [opensearch-project/OpenSearch#15269]
### Is your feature request related to a problem? Please describe

Currently the logic to pick `Ordinals` vs `DirectCollector` is dynamic and based on the number of ordinals to collect and memory overhead. If memory overhead is high for `OrdinalsCollector`, then `DirectCollector` is used. Due to https://issues.apache.org/jira/browse/LUCENE-9663  [few users](https://github.com/opensearch-project/OpenSearch/issues/2820) are reporting regression in Cardinality aggregation because of slower `DirectCollector` after replacing prefix compression with LZ4 compression for terms dictionary in lucene 8.9. 
Fix proposed [here](https://issues.apache.org/jira/browse/LUCENE-10509?focusedCommentId=17519569&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17519569) is to use OrdinalsCollector more often which will collect the ordinals into a bitset first and then performs term lookup in `postCollect()` of segment. 
However, we don't have a way to control picking up `OrdinalsCollector` in OpenSearch. 

### Describe the solution you'd like

Introduce a memory threshold dynamic which OrdinalsCollector can use and if their usage is below it, always use OrdinalsCollector. This can be added [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregator.java#L158). 

### Related component

Search:Aggregations

### Describe alternatives you've considered

Use of eager_global_ordinals or murmur hash, but its an index time setting and can have impact on indexing performance. Its discussed in more detail [here](https://github.com/opensearch-project/OpenSearch/issues/2820#issuecomment-1203403232).

**Comment by mch2** [comment]
Thanks @rishabhmaurya seems like something we should do, am removing untriaged.

**Comment by maitreya2954** [comment]
@mch2 I would like to work on this.

**Comment by jainankitk** [comment]
Thanks @maitreya2954 for volunteering, assigned it to you!

**Comment by rishabhmaurya** [comment]
@maitreya2954 Appreciate you picking this up. Here is what we can do - 

1. Introduce `execution_hint` for cardinality aggregation similar to [term aggregation](https://github.com/opensearch-project/OpenSearch/blob/aca2e9de7daaadee77c33e21ff5215c3f1e8600f/server/src/main/java/org/opensearch/search/aggregations/bucket/terms/TermsAggregationBuilder.java#L75), with values: `direct`, `ordinal` to start with.
2. The default behaviour should be current behaviour. When `execution_hint` is provided, override the collector to be picked here and don't honour this condition [here](https://github.com/opensearch-project/OpenSearch/blob/2f8cb079c2c220304b8e6965853a15732c3bf57c/server/src/main/java/org/opensearch/search/aggregations/metrics/CardinalityAggregator.java#L158). 
3. Add [unit tests](https://github.com/opensearch-project/OpenSearch/blob/2f8cb079c2c220304b8e6965853a15732c3bf57c/server/src/test/java/org/opensearch/search/aggregations/metrics/CardinalityAggregatorTests.java#L87) for all 3 cases. 
4. Add documentation [here](https://github.com/opensearch-project/documentation-website/blob/main/_aggregations/metric/cardinality.md) by opening a documentation issue on this repo. 

If you're blocked don't hesitate to ask for pointers. Thank you.

**Comment by maitreya2954** [comment]
@rishabhmaurya I added execution hint field to the cardinality agg. However, I am stuck at adding unit tests. 

> Add [unit tests](https://github.com/opensearch-project/OpenSearch/blob/2f8cb079c2c220304b8e6965853a15732c3bf57c/server/src/test/java/org/opensearch/search/aggregations/metrics/CardinalityAggregatorTests.java#L87) for all 3 cases.

3 cases being: `direct`, `ordinal` and when no value is provided. right?

Here's my question: What type of query should I test after setting the execution hint. I can see many different type of queries being tested [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/test/java/org/opensearch/search/aggregations/metrics/CardinalityAggregatorTests.java#L87). Should I just do `MatchAllDocsQuery`?

**Comment by rishabhmaurya** [comment]
That's great that you were able to add execution hint. 
> 3 cases being: direct, ordinal and when no value is provided. right?
yes.

> What type of query should I test after setting the execution hint. I can see many different type of queries being tested [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/test/java/org/opensearch/search/aggregations/metrics/CardinalityAggregatorTests.java#L87). Should I just do MatchAllDocsQuery?

You can test for both match_all query and a filter using term query on a field other than cardinality aggregation field.

**Comment by rishabhmaurya** [comment]
@maitreya2954 how were you thinking about checking which collector is picked from tests? 

One way could add a method to this class to get the current collector picked for a given segment and add a logic here to store the name/class - 
https://github.com/opensearch-project/OpenSearch/blob/978d14e867c41b224303b2e77745495336568f79/test/framework/src/main/java/org/opensearch/search/aggregations/AggregatorTestCase.java#L1207

**Comment by maitreya2954** [comment]
@rishabhmaurya Here's the PR: https://github.com/opensearch-project/OpenSearch/pull/15764

**Comment by rishabhmaurya** [comment]
@maitreya2954 great, thanks! I will take a look shortly. Meanwhile you can work on fixing gradle checks which are failing.

**Comment by getsaurabh02** [comment]
Thanks @maitreya2954 for raising the PR. Is it something we can target for 2.18?
cc: @rishabhmaurya

**Comment by maitreya2954** [comment]
@rishabhmaurya From the conversation in the PR, I am assuming we are going to add the cluster setting as well for the memory threshold. Can you guide me to a previous PR that added a cluster setting for reference. 

@getsaurabh02 I am not sure, how long the new changes might take. @rishabhmaurya can we target for 2.18.

**Comment by sandeshkr419** [comment]
@maitreya2954 Are you still working on it?

**Comment by maitreya2954** [comment]
@sandeshkr419 Yes, I will start working was busy with something else. Since, I am very new to this project, can you guide me to a previous PR that added a cluster setting. I will refer that and make changes. Thank you

**Comment by asimmahmood1** [comment]
Plan to follow up to https://github.com/opensearch-project/OpenSearch/pull/17312

**Comment by asimmahmood1** [comment]
So compared 3 different cardinality runs. Current osb cardinality-agg-high uses `agent.name` and there is little difference with default vs ordinals only. `event.id` has almost 7x cardinality and where ordinals is must faster. So I think its worth adding this to osb, e.g. `cardinality-agg-vhigh`

|**osb task**||**cardinality-agg-high**|**cardinality-agg-low**|
|:---|:---|:---|:---|
|field|event.id|agent.name|cloud.region|
|cardinality|180250|26832|25|
|||||
|took_time||||
|no hint|79802|2342|10|
|ordinals|2982|2120|10|
|direct|79673|114579|10|

**Comment by getsaurabh02** [comment]
@asimmahmood1 Do we still need this? Given we have a request level hint now, can we close this?

**Comment by asimmahmood1** [comment]
Proposed change in OSB test case: https://github.com/opensearch-project/opensearch-benchmark-workloads/issues/576

**Comment by asimmahmood1** [comment]
Now that we have the execution_hint out since 2.19.1+, and that its documented in https://opensearch.org/docs/latest/aggregations/metric/cardinality/, we can park this issue.

I think it'll be a better customer experience to include a default setting that uses ordinals more often. Currently, the customer will experience latency regression with data becomes large enough, then research to find out they can use the hint to work around the issue.

**Comment by asimmahmood1** [comment]
Now that hybrid collector has been merged in, I think we can close this one: https://github.com/opensearch-project/OpenSearch/pull/19524
